File: purchaseHandler/BestPurchaseHandler.cpp
//
// Created by Jay on 1/21/2025.
//

#include "BestPurchaseHandler.h"
#include "../game/Game.h"

BestPurchaseHandler::BestPurchaseHandler() {
    std::filesystem::path savedTreesPath = std::filesystem::current_path() / "Saved_Trees";
    std::filesystem::create_directories(savedTreesPath);
    for (const auto& entry : std::filesystem::directory_iterator(savedTreesPath)) {
        if (entry.is_regular_file()) {
            std::string filename = entry.path().filename().string();
            if (filename.size() >= 4 && filename.substr(filename.size() - 4) == ".bin")  {
                // std::cout << filename << std::endl;
                prebuiltKeys[filename] = true;
            }
        }
    }
}

std::pair<uint32_t, buy_list> BestPurchaseHandler::getBestBuy(uint16_t cost, uint16_t weight, const Player& player) {
    const auto total_started = std::chrono::steady_clock::now();
    std::string key = convertWorldToKey(player.player_world);
    auto& key_profile = profiling_per_key[key];
    ++profiling_calls;
    ++key_profile.calls;

    const auto ensure_started = std::chrono::steady_clock::now();
    if (loadedTrees.find(key) == loadedTrees.end()) {
        ++profiling_cache_misses;
        bool loaded = false;
        if (prebuiltKeys.find(key) != prebuiltKeys.end()) {
            try {
                loadTree(key);
                loaded = true;
            } catch (const std::exception&) {
                prebuiltKeys.erase(key);
            }
        }
        if (!loaded) {
            buildTree(player, key);
            saveTree(key);
            prebuiltKeys[key] = true;
        }
    }
    const auto ensure_elapsed = std::chrono::steady_clock::now() - ensure_started;
    profiling_ensure_time += ensure_elapsed;
    key_profile.ensure_time += ensure_elapsed;

    const auto query_started = std::chrono::steady_clock::now();
    const auto result = loadedTrees[key].query(cost, weight);
    const auto query_elapsed = std::chrono::steady_clock::now() - query_started;
    profiling_query_time += query_elapsed;
    key_profile.query_time += query_elapsed;

    const auto total_elapsed = std::chrono::steady_clock::now() - total_started;
    profiling_total_time += total_elapsed;
    key_profile.total_time += total_elapsed;
    return result;
}

void BestPurchaseHandler::prebuildSupportedTrees() {
    for (const Player& player : buildSupportedWorldPlayers()) {
        const std::string key = convertWorldToKey(player.player_world);
        if (loadedTrees.find(key) == loadedTrees.end()) {
            bool loaded = false;
            if (prebuiltKeys.find(key) != prebuiltKeys.end()) {
                try {
                    loadTree(key);
                    loaded = true;
                } catch (const std::exception&) {
                    prebuiltKeys.erase(key);
                }
            }
            if (!loaded) {
                buildTree(player, key);
                saveTree(key);
                prebuiltKeys[key] = true;
            }
        }
    }
}

void BestPurchaseHandler::warmSupportedTrees() {
    for (const Player& player : buildSupportedWorldPlayers()) {
        const std::string key = convertWorldToKey(player.player_world);
        if (loadedTrees.find(key) != loadedTrees.end()) {
            std::cout << "[*] Warm tree already loaded: " << key << '\n';
            continue;
        }

        std::cout << "[*] Warming tree: " << key << '\n';
        bool loaded = false;
        if (prebuiltKeys.find(key) != prebuiltKeys.end()) {
            try {
                loadTree(key);
                loaded = true;
                std::cout << "[*] Loaded tree from disk: " << key << '\n';
            } catch (const std::exception&) {
                prebuiltKeys.erase(key);
            }
        }

        if (!loaded) {
            std::cout << "[*] Building missing warm tree: " << key << '\n';
            buildTree(player, key);
            saveTree(key);
            prebuiltKeys[key] = true;
            std::cout << "[*] Built and saved warm tree: " << key << '\n';
        }
    }
}

bool BestPurchaseHandler::verifySupportedTrees(const uint16_t max_gold, const uint16_t max_weight, std::ostream& out) {
    bool all_ok = true;
    for (const Player& player : buildSupportedWorldPlayers()) {
        const std::string key = convertWorldToKey(player.player_world);
        if (loadedTrees.find(key) == loadedTrees.end()) {
            bool loaded = false;
            if (prebuiltKeys.find(key) != prebuiltKeys.end()) {
                try {
                    loadTree(key);
                    loaded = true;
                } catch (const std::exception&) {
                    prebuiltKeys.erase(key);
                }
            }
            if (!loaded) {
                buildTree(player, key);
                saveTree(key);
                prebuiltKeys[key] = true;
            }
        }

        out << "[*] Verifying " << key << " up to gold=" << max_gold << " weight=" << max_weight << '\n';
        for (uint16_t gold = 0; gold <= max_gold; ++gold) {
            for (uint16_t weight = 0; weight <= max_weight; ++weight) {
                const auto fenwick = loadedTrees[key].query(gold, weight);
                const auto brute = bruteForceBestBuy(player, gold, weight);
                const uint16_t fenwick_cost = calculateCost(player, fenwick.second);
                const uint16_t fenwick_weight = calculateWeight(player, fenwick.second);
                const uint32_t fenwick_profit = calculateProfit(player, fenwick.second);
                if (fenwick_cost > gold || fenwick_weight > weight || fenwick_profit != fenwick.first || fenwick.first != brute.first) {
                    out << "[!] Mismatch for " << key
                        << " gold=" << gold
                        << " weight=" << weight
                        << " fenwick_profit=" << fenwick.first
                        << " brute_profit=" << brute.first
                        << " fenwick_cost=" << fenwick_cost
                        << " fenwick_weight=" << fenwick_weight
                        << " realized_profit=" << fenwick_profit << '\n';
                    all_ok = false;
                    return false;
                }
            }
        }
    }
    return all_ok;
}

void BestPurchaseHandler::loadTree(const std::string& key) {
    loadedTrees[key] = FenwickTree2D();
    try {
        loadedTrees[key].deserialize(key);
    } catch (const std::exception& e) {
        throw std::runtime_error("Deserialization failed for tree: " + key + ", Error: " + e.what());
    }
}

void BestPurchaseHandler::saveTree(const std::string& key) {
    loadedTrees[key].serialize(key);
}

void BestPurchaseHandler::buildTree(const Player& player, const std::string& key) {
    std::cout << "[*] Building Tree " << key << std::endl;
    std::vector<buy_option> elements;
    generateElementsForWorld(player, elements);
    loadedTrees[key] = FenwickTree2D(elements);
    std::cout << "[*] Build Finished" << std::endl;
}

std::vector<Player> BestPurchaseHandler::buildSupportedWorldPlayers() {
    std::vector<Player> players;
    players.reserve(std::size(VALID_TOWNS) * std::size(VALID_MERCHANTS) * std::size(VALID_WITCHES));

    for (const auto town : VALID_TOWNS) {
        Player base_player;
        Game::apply_town(base_player, town);

        for (const auto merch : VALID_MERCHANTS) {
            Player merch_player = base_player;
            if (merch != m_No_Merchant) {
                merch_player.gold = std::max<int>(merch_player.gold, 200);
                if (!Game::event_merchant(merch_player, merch)) {
                    continue;
                }
                merch_player.day_start_skip = false;
            }

            for (const auto witch : VALID_WITCHES) {
                Player world_player = merch_player;
                if (witch != w_No_Witch) {
                    Game::event_witch(world_player, witch);
                    world_player.day_start_skip = false;
                }
                players.emplace_back(std::move(world_player));
            }
        }
    }
    return players;
}

std::pair<uint32_t, buy_list> BestPurchaseHandler::bruteForceBestBuy(const Player& player, const uint16_t cost_limit, const uint16_t weight_limit) {
    uint32_t best_profit = 0;
    buy_list best_buy;

    const int max_spice = player.item_shop[i_Spice][id_Buy] == 0 ? weight_limit : std::min<int>(cost_limit / player.item_shop[i_Spice][id_Buy], weight_limit / std::max(player.item_shop[i_Spice][id_Weight], 1));
    const int max_pottery = player.item_shop[i_Pottery][id_Buy] == 0 ? weight_limit : std::min<int>(cost_limit / player.item_shop[i_Pottery][id_Buy], weight_limit / std::max(player.item_shop[i_Pottery][id_Weight], 1));
    const int max_marble = player.item_shop[i_Marble][id_Buy] == 0 ? weight_limit : std::min<int>(cost_limit / player.item_shop[i_Marble][id_Buy], weight_limit / std::max(player.item_shop[i_Marble][id_Weight], 1));
    const int max_silk = player.item_shop[i_Silk][id_Buy] == 0 ? weight_limit : std::min<int>(cost_limit / player.item_shop[i_Silk][id_Buy], weight_limit / std::max(player.item_shop[i_Silk][id_Weight], 1));
    const int max_jewelry = player.item_shop[i_Jewelry][id_Buy] == 0 ? weight_limit : std::min<int>(cost_limit / player.item_shop[i_Jewelry][id_Buy], weight_limit / std::max(player.item_shop[i_Jewelry][id_Weight], 1));

    for (int spice = 0; spice <= max_spice; ++spice) {
        for (int pottery = 0; pottery <= max_pottery; ++pottery) {
            for (int marble = 0; marble <= max_marble; ++marble) {
                for (int silk = 0; silk <= max_silk; ++silk) {
                    for (int jewelry = 0; jewelry <= max_jewelry; ++jewelry) {
                        const int cost =
                            spice * player.item_shop[i_Spice][id_Buy] +
                            pottery * player.item_shop[i_Pottery][id_Buy] +
                            marble * player.item_shop[i_Marble][id_Buy] +
                            silk * player.item_shop[i_Silk][id_Buy] +
                            jewelry * player.item_shop[i_Jewelry][id_Buy];
                        const int weight =
                            spice * player.item_shop[i_Spice][id_Weight] +
                            pottery * player.item_shop[i_Pottery][id_Weight] +
                            marble * player.item_shop[i_Marble][id_Weight] +
                            silk * player.item_shop[i_Silk][id_Weight] +
                            jewelry * player.item_shop[i_Jewelry][id_Weight];

                        if (cost > cost_limit || weight > weight_limit) {
                            continue;
                        }

                        const uint32_t profit =
                            spice * std::max(player.item_shop[i_Spice][id_Sell] - player.item_shop[i_Spice][id_Buy], 0) +
                            pottery * std::max(player.item_shop[i_Pottery][id_Sell] - player.item_shop[i_Pottery][id_Buy], 0) +
                            marble * std::max(player.item_shop[i_Marble][id_Sell] - player.item_shop[i_Marble][id_Buy], 0) +
                            silk * std::max(player.item_shop[i_Silk][id_Sell] - player.item_shop[i_Silk][id_Buy], 0) +
                            jewelry * std::max(player.item_shop[i_Jewelry][id_Sell] - player.item_shop[i_Jewelry][id_Buy], 0);

                        if (profit > best_profit) {
                            best_profit = profit;
                            best_buy = buy_list(0, spice, pottery, marble, silk, jewelry);
                        }
                    }
                }
            }
        }
    }

    return {best_profit, best_buy};
}

uint32_t BestPurchaseHandler::calculateProfit(const Player& player, const buy_list& buy) {
    return
        buy.spice * std::max(player.item_shop[i_Spice][id_Sell] - player.item_shop[i_Spice][id_Buy], 0) +
        buy.pottery * std::max(player.item_shop[i_Pottery][id_Sell] - player.item_shop[i_Pottery][id_Buy], 0) +
        buy.marble * std::max(player.item_shop[i_Marble][id_Sell] - player.item_shop[i_Marble][id_Buy], 0) +
        buy.silk * std::max(player.item_shop[i_Silk][id_Sell] - player.item_shop[i_Silk][id_Buy], 0) +
        buy.jewelry * std::max(player.item_shop[i_Jewelry][id_Sell] - player.item_shop[i_Jewelry][id_Buy], 0);
}

uint16_t BestPurchaseHandler::calculateCost(const Player& player, const buy_list& buy) {
    return static_cast<uint16_t>(
        buy.spice * player.item_shop[i_Spice][id_Buy] +
        buy.pottery * player.item_shop[i_Pottery][id_Buy] +
        buy.marble * player.item_shop[i_Marble][id_Buy] +
        buy.silk * player.item_shop[i_Silk][id_Buy] +
        buy.jewelry * player.item_shop[i_Jewelry][id_Buy]
    );
}

uint16_t BestPurchaseHandler::calculateWeight(const Player& player, const buy_list& buy) {
    return static_cast<uint16_t>(
        buy.spice * player.item_shop[i_Spice][id_Weight] +
        buy.pottery * player.item_shop[i_Pottery][id_Weight] +
        buy.marble * player.item_shop[i_Marble][id_Weight] +
        buy.silk * player.item_shop[i_Silk][id_Weight] +
        buy.jewelry * player.item_shop[i_Jewelry][id_Weight]
    );
}

BestPurchaseHandler::ProfilingSummary BestPurchaseHandler::getProfilingSummary() const {
    const auto to_ms = [](const std::chrono::nanoseconds duration) {
        return std::chrono::duration<double, std::milli>(duration).count();
    };

    ProfilingSummary summary;
    summary.calls = profiling_calls;
    summary.cache_misses = profiling_cache_misses;
    summary.total_ms = to_ms(profiling_total_time);
    summary.query_ms = to_ms(profiling_query_time);
    summary.ensure_ms = to_ms(profiling_ensure_time);

    double slowest_total_ms = -1.0;
    double slowest_query_ms = -1.0;
    for (const auto& [key, data] : profiling_per_key) {
        const double total_ms = to_ms(data.total_time);
        const double query_ms = to_ms(data.query_time);
        const double ensure_ms = to_ms(data.ensure_time);
        if (total_ms > slowest_total_ms) {
            slowest_total_ms = total_ms;
            summary.slowest_total_key = ProfilingKeySummary{key, data.calls, total_ms, query_ms, ensure_ms};
        }
        if (query_ms > slowest_query_ms) {
            slowest_query_ms = query_ms;
            summary.slowest_query_key = ProfilingKeySummary{key, data.calls, total_ms, query_ms, ensure_ms};
        }
    }

    return summary;
}

File: purchaseHandler/best_purchase_handler.py
from ..consts.Enums import TownOptions, MerchOptions, WitchOptions, Items, ItemData
from ..player.player import Player
from fenwick_tree_2d import FenwickTree2D
from math import ceil
from pathlib import Path

MAX_WEIGHT = 360 # This is from internal tests but could be a little higher
MAX_GOLD = 50000 #

valid_towns = [
    TownOptions.Normalia,    # Default
    TownOptions.Mahaji,      # Spice +5 Sell
    TownOptions.Huangoz,     # Marble -20 Buy
    TownOptions.Cornecopea,  # Food is free
    TownOptions.Gayawaku,    # Food is weightless
    TownOptions.Ipartus      # Silk -100 Buy
]

valid_merchant = [
    MerchOptions.No_Merchant,     # Merch has not happened or can not afford
    MerchOptions.Route_to_Mahaji  # Spice +20 Sell
]

valid_witch = [
    WitchOptions.No_Witch,            # Witch has not happened
    WitchOptions.Vertue_of_Patience,  # Jewelry +100 Buy
    WitchOptions.Midas_was_a_Trader,  # Marble -300 Sell
    WitchOptions.Time_is_Money,       # Jewelry -500 Sell
    WitchOptions.Oasis_of_Sanctifan,  # Silk +200 Buy
    WitchOptions.The_Stonecutter      # Marble Weight -15 and Buy +60
]


def convert_world_to_key(world: str) -> str:
    town, merch, witch = [int(x) for x in world.split('_')]
    if town not in valid_towns:
        town = TownOptions.Normalia
    if merch not in valid_merchant:
        merch = MerchOptions.No_Merchant
    if witch not in valid_witch:
        witch = WitchOptions.No_Witch
    return f'tree_{town}_{merch}_{witch}.pkl'


def generate_elements_for_world(player: Player) -> list[list[int | list[int]]]:
    max_spice = ceil(MAX_WEIGHT / player.item_shop[Items.Spice][ItemData.weight])
    max_pottery = ceil(MAX_WEIGHT / player.item_shop[Items.Pottery][ItemData.weight])
    max_marble = ceil(MAX_WEIGHT / player.item_shop[Items.Marble][ItemData.weight])
    max_silk = ceil(MAX_WEIGHT / player.item_shop[Items.Silk][ItemData.weight])
    max_jewelry = ceil(MAX_WEIGHT / player.item_shop[Items.Jewelry][ItemData.weight])
    found = {}
    found_list = []
    for s in range(max_spice):
        for p in range(max_pottery):
            for m in range(max_marble):
                for i in range(max_silk):
                    for j in range(max_jewelry):
                        profit = ((player.item_shop[Items.Spice][ItemData.sell] - player.item_shop[Items.Spice][ItemData.buy]) * s +
                                 (player.item_shop[Items.Pottery][ItemData.sell] - player.item_shop[Items.Pottery][ItemData.buy]) * p +
                                 (player.item_shop[Items.Marble][ItemData.sell] - player.item_shop[Items.Marble][ItemData.buy]) * m +
                                 (player.item_shop[Items.Silk][ItemData.sell] - player.item_shop[Items.Silk][ItemData.buy]) * i +
                                 (player.item_shop[Items.Jewelry][ItemData.sell] - player.item_shop[Items.Jewelry][ItemData.buy]) * j)
                        weight = (player.item_shop[Items.Spice][ItemData.weight] * s +
                                 player.item_shop[Items.Pottery][ItemData.weight] * p +
                                 player.item_shop[Items.Marble][ItemData.weight] * m +
                                 player.item_shop[Items.Silk][ItemData.weight] * i +
                                 player.item_shop[Items.Jewelry][ItemData.weight] * j)
                        cost = (player.item_shop[Items.Spice][ItemData.buy] * s +
                               player.item_shop[Items.Pottery][ItemData.buy] * p +
                               player.item_shop[Items.Marble][ItemData.buy] * m +
                               player.item_shop[Items.Silk][ItemData.buy] * i +
                               player.item_shop[Items.Jewelry][ItemData.buy] * j)
                        index = cost + weight * 1j
                        if (MAX_GOLD < cost and MAX_WEIGHT < weight) or profit == 0:
                            continue
                        if index not in found:
                            found[index] = len(found_list)
                            found_list.append([cost, weight, profit, [0, s, p, m, i, j]])
                        elif found_list[found[index]][2] < profit:
                            found_list[found[index]][2] = profit
                            found_list[found[index]][3] = [0, s, p, m, i, j]
    return found_list


class BestPurchaseHandler:
    def __init__(self):
        self.prebuilt_keys = {}
        self.loaded_trees = {}

        saved_trees_path = Path.cwd() / "Saved_Trees"
        if saved_trees_path.exists():
            for entry in saved_trees_path.iterdir():
                if entry.is_file() and entry.suffix == ".pkl":
                    filename = entry.name
                    self.prebuilt_keys[filename] = True

    def get_best_buy(self, cost: int, weight: int, player: Player) -> tuple[int, list[int]]:
        key = convert_world_to_key(player.world)

        if key not in self.loaded_trees:
            if key in self.prebuilt_keys:
                self.load_tree(key)
            else:
                self.build_tree(player, key)
                self.save_tree(key)

        # Query the tree
        return self.loaded_trees[key].query(cost, weight)

    def load_tree(self, key: str) -> None:
        self.loaded_trees[key] = FenwickTree2D([])
        try:
            self.loaded_trees[key].deserialize(key)
        except Exception as e:
            raise RuntimeError(f"Deserialization failed for tree: {key}, Error: {str(e)}")

    def save_tree(self, key: str) -> None:
        self.loaded_trees[key].serialize(key)

    def build_tree(self, player: Player, key: str) -> None:
        self.loaded_trees[key] = FenwickTree2D(generate_elements_for_world(player))